providers/siw/siw.c

```diff
@@ -0,0 +1,941 @@
+// SPDX-License-Identifier: GPL-2.0 or BSD-3-Clause
+
+// Authors: Bernard Metzler <bmt@zurich.ibm.com>
+// Copyright (c) 2008-2019, IBM Corporation
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/mman.h>
+#include <net/if.h>
+#include <pthread.h>
+#include <stdatomic.h>
+
+#include "siw.h"
+#include "siw_abi.h"
+
+static int siw_debug;
+
+static int siw_query_device(struct ibv_context *ctx,
+			    struct ibv_device_attr *attr)
+{
+	struct ibv_query_device cmd;
+	uint64_t raw_fw_ver;
+	unsigned int major, minor, sub_minor;
+	int rv;
+
+	memset(&cmd, 0, sizeof(cmd));
+
+	rv = ibv_cmd_query_device(ctx, attr, &raw_fw_ver, &cmd, sizeof(cmd));
```

Comment:
No, all udatas have to be in structs and declared in the various abi files. Do not pass loose values like this through udata, it makes a big mess to figure out.

Reply:
Can you please help me here - I do not see what is wrong with it
(it is basically rather the same as all the other provides query_device()
function)?

Reply:
Grr.. the other providers are wrong too, I thought I got rid of all this stuff :(

All buffers passed to these ibv_cmd functions by drivers must be formed by the DECLARE_DRV_CMD() macros.

Reply:
OK I'll change it accordingly.

Reply:
Oh, my remarks here are not right, raw_fw_ver is just some hacky thing that got added, not a udata like in every other command.. Never mind then.